src/array/cpu/spmm.h

```diff
@@ -12,15 +16,568 @@
+  printf("stage1: %ld ticks\n", endTick - startTick);
+#endif
+
+#if SPMM_LOG_INFO
+  startTick = __rdtsc();
+#endif
+  int _ld = N;
+  libxsmm_meltwfunction_opreduce_vecs_idx kernel = nullptr;
+  kernel = SpMMCreateLibxsmmKernel<IdType, DType, Op>(has_idx, N, _ld, LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_REDOP_SUM, false);
+#if SPMM_LOG_INFO
+  endTick = __rdtsc();
+  printf("stage2: %ld ticks\n", endTick - startTick);
+
+  startTick = __rdtsc();
+#endif
+  SpMMBlockwiseOpSum(block_csr_array, B, E, C, has_idx, N, num_M_blocks, num_K_blocks, M_BLOCK_SIZE, kernel);
+#if SPMM_LOG_INFO
+  endTick = __rdtsc();
+#endif
+#if SPMM_LOG_INFO
+  printf("stage3: %ld ticks\n", endTick - startTick);
+  startTick = __rdtsc();
+#endif
+
+  SpMMFreeBlocks(block_csr_array, num_M_blocks, num_K_blocks, Op::use_lhs, Op::use_rhs);
+#if SPMM_LOG_INFO
+  endTick = __rdtsc();
+  printf("stage4: %ld ticks\n", endTick - startTick);
+#endif
+}
+ 
+template <typename IdType, typename DType, typename Op, typename Cmp>
+void SpMMCmpCsrOpt(
+    const BcastOff& bcast,
+    const CSRMatrix& csr,
+    NDArray ufeat, NDArray efeat,
+    NDArray out,
+    NDArray argu, NDArray arge) {
+
+  int32_t LLC_SIZE = get_llc_size();
+
+#if SPMM_LOG_INFO
+  uint64_t startTick, endTick;
+  startTick = __rdtsc();
+#endif
+
+  const bool has_idx = !IsNullArray(csr.data);
+
+  DType* C = out.Ptr<DType>();
+  DType* B = ufeat.Ptr<DType>();
+  DType* E = efeat.Ptr<DType>();
+  IdType *argB = argu.Ptr<IdType>();
+  IdType *argE = arge.Ptr<IdType>();
+
+  int nthreads = omp_get_max_threads();
+  const IdType M = csr.num_rows;
+  const IdType N = bcast.out_len;
+  const IdType K = csr.num_cols;
+  IdType* IndPtr = csr.indptr.Ptr<IdType>();
+  assert(IndPtr != nullptr);
+  int total_nnz = IndPtr[M];
+  if(M <= 0 || K <= 0 || N <= 0 || total_nnz <= 0) return;
+
+  float avgDegree = total_nnz * 1.0 / M;
+  float nnz_prob = avgDegree / K;
+
+  IdType K_BLOCK_SIZE = LLC_SIZE / (N * sizeof(DType) * nnz_prob * 500);
+  IdType M_BLOCK_SIZE = M / (nthreads * 20);
+  if (M_BLOCK_SIZE == 0) M_BLOCK_SIZE = 1;
+  if (K_BLOCK_SIZE == 0) K_BLOCK_SIZE = 1;
+
+  IdType num_M_blocks = (M + M_BLOCK_SIZE - 1) / M_BLOCK_SIZE;
+  IdType num_K_blocks = (K + K_BLOCK_SIZE - 1) / K_BLOCK_SIZE;
+
+  CSRM<IdType, IdType> *block_csr_array = (CSRM<IdType, IdType> *)aligned_alloc(64, sizeof(CSRM<IdType, IdType>) * num_M_blocks * num_K_blocks);
+#if SPMM_LOG_INFO
+  endTick = __rdtsc();
+  printf("spmmcmpcsr nthreads = %d, LLC_SIZE = %d\n", nthreads, LLC_SIZE);
+  printf("M = %d, K = %d, N = %d, use_lhs = %d, use_rhs = %d\n", M, K, N, Op::use_lhs, Op::use_rhs);
+  printf("total_nnz = %d, avgDegree = %f\n", total_nnz, avgDegree);
+  printf("has_idx = %d\n", has_idx);
+  printf("nnz_prob = %lf\n", nnz_prob);
+  printf("K_BLOCK_SIZE = %d, M_BLOCK_SIZE = %d\n", K_BLOCK_SIZE, M_BLOCK_SIZE);
+  printf("num_K_blocks = %d, num_M_blocks = %d\n", num_K_blocks, num_M_blocks);
+  printf("stage0: %ld ticks\n", endTick - startTick);
```

Comment:
here it prints too much information.

Reply:
All of this is disabled by a macro (SPMM_LOG_INFO) and will be enabled only for debugging.

Reply:
Could you use DGL's log macro for code consistency?

Reply:
Sure! What is the preferred way to add debugging logs? Should I use DLOG(INFO) or use:
#ifdef DEBUG
LOG(INFO) << ...;
#endif